### PR TITLE
Adding HAProxy module

### PR DESCRIPTION
**Pull request by hartfordfive**
PR for issue #2247

### CODE REVIEW

File: metricbeat/logs/metricbeat [hunk 1]
```diff
@@ -0,0 +1,10 @@
+2016-08-10T19:06:01-04:00 INFO Metrics logging every 30s
```
Comment (ruflin) [inline metricbeat/logs/metricbeat]: The logs directory shuld not be commited. it is under .gitignore in the master branch

File: metricbeat/module/haproxy/haproxy.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+		return nil, err
+	}
+
+	if b, err := ioutil.ReadAll(res); err == nil {
+
+		resultMap := map[string]interface{}{}
+
+		for _, ln := range strings.Split(string(b), "\n") {
+
+			ln := strings.Trim(ln, " ")
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: There is the [`strings.TrimSpace`](https://golang.org/pkg/strings/#TrimSpace) function for this.

File: metricbeat/module/haproxy/haproxy.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+			if ln == "" {
+				continue
+			}
+
+			parts := strings.Split(strings.Trim(ln, " "), ":")
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: `ln` has already been trimmed at this point.

File: metricbeat/module/haproxy/haproxy.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+				continue
+			}
+
+			resultMap[parts[0]] = strings.Trim(parts[1], " ")
+		}
+
+		var result *Info
+		err := mapstructure.Decode(resultMap, &result)
+		if err != nil {
+			panic(err)
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Return the error rather than `panic`'ing.

File: metricbeat/module/haproxy/haproxy.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing CSV: %s", err)
+	}
+
+	return statRes, nil
+
+}
+
+// GetInfo returns the result from the 'show stat' command
+func (c *Client) GetInfo() (infoRes *Info, err error) {
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Don't use named return parameters unless you have good reason to. This applies to other parts of the code. https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: metricbeat/module/haproxy/info/_meta/fields.yml [hunk 6]
```diff
@@ -0,0 +1,194 @@
+- name: info
+  type: group
+  description: >
+    General infomration collected on HAProxy process
+  fields:
+    - name: nb_proc
+      type: intger
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/_meta/fields.yml]: `s/intger/integer/`

But instead of integer use `long`.

File: metricbeat/module/haproxy/info/_meta/fields.yml [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    - name: tasks
+      type: intger
+      description: >
+
+    - name: run_queue
+      type: intger
+      description: >
+
+    - name: idle_pct
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/_meta/fields.yml]: Consider making this value range on [0, 1] so you can use a float with a percentage formatter. For example: https://github.com/elastic/beats/blob/master/metricbeat/module/system/memory/_meta/fields.yml#L26-L27

File: metricbeat/module/haproxy/info/info.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch() (common.MapStr, error) {
+
+	hapc, err := haproxy.NewHaproxyClient(m.statsAddr)
+	if err != nil {
+		return nil, fmt.Errorf(fmt.Sprintf("HAProxy Client error: %s", err))
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/info.go]: `fmt.Errorf()` accepts the same parameters as `fmt.Sprintf` so you can just drop the call to `Sprintf`.

File: metricbeat/module/haproxy/haproxy.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+// Run sends a designated command to the haproxy stats socket
+func (c *Client) run(cmd string) (*bytes.Buffer, error) {
+	var conn net.Conn
+	response := bytes.NewBuffer(nil)
+
+	conn, err := net.Dial(c.ProtoScheme, c.Address)
+	if err != nil {
+		return response, err
+	}
+	c.connection = conn
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Why store the connection in the `Client` if it's not going to be reused?

**Comment by andrewkroh** [comment]
@hartfordfive Looks like you are off to a great start. Thanks for the contribution.

File: metricbeat/module/haproxy/haproxy.go (hunk 9 above)
Comment (hartfordfive) [inline metricbeat/module/haproxy/haproxy.go]: Good point.  I've changed that.

**Comment by hartfordfive** [comment]
@andrewkroh I think I've fixed most of the issues you've commented on although I have a few other HAProxy stats related bugs, so I'll let you know once that's done and I'm ready for a final review.

**Comment by hartfordfive** [comment]
@andrewkroh I think the PR should be good for another review when ever you get the chance.

**Comment by ruflin** [comment]
@hartfordfive Any chance you could rebase the commits or merge in master so the merging is possible? Just to make we review the most recent changes on top of master. It seems that you merged in master so rebasing could become ugly. TBH not sure what the best recommendation is here as you have lots of commits. Probably best keep merging in master and we will to the squash and rebase on our side at the end.

File: metricbeat/.gitignore [hunk 10]
```diff
@@ -3,3 +3,4 @@ build
 /metricbeat
 /metricbeat.test
 /docs/html_docs
+/logs/*
```
Comment (ruflin) [inline metricbeat/.gitignore]: This should be not needed anymore as soon as this is rebased on top of master or master is merged.

File: metricbeat/logs/metricbeat [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+2016-08-10T19:06:01-04:00 INFO Loading template enabled. Reading template file: /Users/alain.lefebvre/Documents/code/github_personal/beats/metricbeat/metricbeat.template.json
```
Comment (ruflin) [inline metricbeat/logs/metricbeat]: This file must be removed manually as it is already commited.

File: metricbeat/metricbeat.yml [hunk 12]
```diff
@@ -36,13 +36,24 @@ metricbeat.modules:
 
     # Per process stats
     - process
-  enabled: true
+  enabled: false
```
Comment (ruflin) [inline metricbeat/metricbeat.yml]: This change should be reverted

File: metricbeat/metricbeat.yml [hunk 13]
```diff
@@ -36,13 +36,24 @@ metricbeat.modules:
...
 
   # if true, exports the CPU usage in ticks, together with the percentage values
   cpu_ticks: false
 
+#------------------------------- haproxy Module ------------------------------
+- module: haproxy
+  metricsets: 
+    - "stat"
+    - "info"
+  enabled: true
```
Comment (ruflin) [inline metricbeat/metricbeat.yml]: haproxy should be disabled by default in the short config

File: metricbeat/module/haproxy/_meta/haproxy.conf [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+  maxconn 4000
+  #daemon
+  user  root
+  group staff
+
+  # for restarts
+  pidfile /var/run/haproxy.pid
+  # Logging to syslog facility local0
+  log   127.0.0.1       local0
+  stats socket 0.0.0.0:14567
```
Comment (ruflin) [inline metricbeat/module/haproxy/_meta/haproxy.conf]: Is this something that is enabled by default in most haproxy setups? If not, could be worth to mention this in the haproxy module docs.

File: metricbeat/module/haproxy/info/_meta/data.json [hunk 15]
```diff
@@ -0,0 +1,59 @@
+{
+  "@timestamp": "2016-08-13T19:44:21.036Z",
+  "beat": {
+    "hostname": "beathost",
+    "name": "beathost"
+  },
+  "haproxy": {
+    "info": {
+      "compress_bps_in": 0,
```
Comment (ruflin) [inline metricbeat/module/haproxy/info/_meta/data.json]: The event conventions should be applied here: https://www.elastic.co/guide/en/beats/libbeat/5.0/event-conventions.html

**Comment by ruflin** [comment]
@hartfordfive @andrewkroh I suggest we mark the haproxy metricsets as experimental first. We do this at first for all new metricsets to make it possible to the metricsets in early and test them, but still change the data structure if needed.

@hartfordfive Making it experimental only means adding a log message to the Metricset and a flag in the docs. Have a look at filebeat metricset as an example.

**Comment by hartfordfive** [comment]
@ruflin @andrewkroh I agree it would be best to flag this module as experimental for while before it's considered stable.  I'll apply those other changes you mentioned above shortly.

File: metricbeat/module/haproxy/_meta/docs.asciidoc [hunk 16]
```diff
@@ -0,0 +1,9 @@
+== haproxy Module
+
+This is the haproxy Module.  To enable stats collection from HAProxy, you must enable the stats socket via TCP.  
+For example, placing the following statement under the `global` or `default` section of the haproxy config:
+
+`stats socket 0.0.0.0:14567`
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/_meta/docs.asciidoc]: I recommend providing a more secure example. Either show an example that binds only to `127.0.0.1` or use a unix socket.

File: metricbeat/module/haproxy/haproxy.go [hunk 17]
```diff
@@ -0,0 +1,243 @@
+package haproxy
+
+import (
+	"bytes"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"github.com/gocarina/gocsv"
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Please separate the stdlib imports from the third-party imports. Most of our code follows this style if you would like to see some examples. Checkout the [goimports](http://goinbigdata.com/goimports-vs-gofmt/#goimports) tool.

File: metricbeat/module/haproxy/haproxy.go [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+
+// NewHaproxyClient returns a new instance of HaproxyClient
+func NewHaproxyClient(address string) (*Client, error) {
+	parts := strings.Split(address, "://")
+	if len(parts) != 2 {
+		return nil, errors.New("Must have protocol scheme and address!")
+	}
+
+	if parts[0] != "tcp" && parts[0] != "unix" {
+		return nil, errors.New("Invalid Protocol Scheme!")
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Error strings should not be capitalized. https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: metricbeat/module/haproxy/haproxy.go [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+			if len(parts) != 2 {
+				continue
+			}
+
+			resultMap[parts[0]] = strings.TrimSpace(parts[1])
+		}
+
+		var result *Info
+		err := mapstructure.Decode(resultMap, &result)
+		if err != nil {
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/haproxy.go]: Consider writing this which is slightly more concise. (not a mandatory change, it's just how I'd of written it)

```
if err := mapstructure.Decode(resultMap, &result); err != nil {
    return nil, err
}
```

File: metricbeat/module/haproxy/info/_meta/fields.yml [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+        Process ID
+
+    - name: uptime_sec
+      type: integer
+      description: >
+        Current uptime in seconds
+
+    - name: mem_max_mb
+      type: integer
+      format: bytes
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/_meta/fields.yml]: If the value is actually in MB then you don't want to format it as `bytes`. I recommend converting the value to bytes and renaming the field so you can take advantage of the byte formatter in Kibana (it uses http://numeraljs.com/).

Comment (andrewkroh) [inline metricbeat/module/haproxy/info/_meta/fields.yml]: Oh, I see you do convert the value. So then I think the field just needs a rename to reflect the correct units.

File: metricbeat/module/haproxy/info/data.go [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+	source = map[string]interface{}{}
+
+	for i := 0; i < st.NumField(); i++ {
+		f := st.Field(i)
+
+		if typeOfT.Field(i).Name == "IdlePct" {
+			// Convert this value to a float between 0.0 and 1.0
+			fval, err := strconv.ParseFloat(f.Interface().(string), 64)
+			if err != nil {
+				panic(err)
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/data.go]: I don't recommend panic'ing. (https://github.com/golang/go/wiki/CodeReviewComments#dont-panic) Instead, either propagate the error or handle it.

File: metricbeat/module/haproxy/stat/data.go [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+	}
+	return results
+}
+
+// Map data to MapStr
+func eventMapping(info []*haproxy.Stat) []common.MapStr {
+
+	var events []common.MapStr
+	source := map[string]interface{}{}
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/stat/data.go]: Doesn't look like this allocated map is ever used. You can remove it and declare `source` within the loop.

File: metricbeat/module/haproxy/info/data.go [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+
+// Map data to MapStr
+func eventMapping(info *haproxy.Info) common.MapStr {
+	// Full mapping from info
+
+	source := map[string]interface{}{}
+
+	st := reflect.ValueOf(info).Elem()
+	typeOfT := st.Type()
+	source = map[string]interface{}{}
```
Comment (andrewkroh) [inline metricbeat/module/haproxy/info/data.go]: This is allocating a second map. I don't see a reason for it given the `source` was allocated above.

**Comment by andrewkroh** [comment]
I looked over the Golang code and left a few comments. 

I didn't review the field names very closely to check if they conform to the [event conventions](https://www.elastic.co/guide/en/beats/libbeat/5.0/event-conventions.html), nor did I review the field documentation closely (maybe @dedemorton want to look at the docs, or maybe she prefers to edit them after they are merged).

**Comment by dedemorton** [comment]
It's easier for me to edit the doc changes after they are merged. Please make sure you do a local doc build before the changes are merged to ensure that your contribution doesn't break the doc build. Many thanks!

File: metricbeat/module/haproxy/haproxy.go (hunk 19 above)
Comment (hartfordfive) [inline metricbeat/module/haproxy/haproxy.go]: Yes I definitely agree it's cleaner that way, I'll definitely change that.

**Comment by hartfordfive** [comment]
I've applied most of the modifications for the comments you mentioned above @andrewkroh although i still have to look into the event conventions.  @dedemorton as for the local doc build, where can I see some details regarding that procedure?

**Comment by andrewkroh** [comment]
@hartfordfive  Can you fix the merge conflicts. You probably need to rebase and address a few conflicts.

Once updated I can test it on Jenkins which will also do a doc build so you don't have to. But if you want to test locally, the command is `make docs`.

**Comment by hartfordfive** [comment]
@andrewkroh I actually merged in the master and fixed the merge conflicts a few days ago. Should I do it again?

**Comment by hartfordfive** [comment]
@andrewkroh Ok I've done a few other changes and also updated the index templates to follow the updated event conventions.  Let me know how it looks.

File: metricbeat/etc/beat.yml [hunk 24]
```diff
@@ -43,4 +43,6 @@ metricbeat.modules:
   period: 10s
   processes: ['.*']
 
+  # if true, exports the CPU usage in ticks, together with the percentage values
```
Comment (ruflin) [inline metricbeat/etc/beat.yml]: Not sure how this change neded up in this PR?

File: vendor/vendor.json [hunk 25]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (ruflin) [inline vendor/vendor.json]: Which tool creates this file?

File: vendor/github.com/gocarina/gocsv/README.md [hunk 26]
```diff
@@ -0,0 +1,151 @@
+Go CSV
```
Comment (ruflin) [inline vendor/github.com/gocarina/gocsv/README.md]: So far we added a vendor directory in the module directly if the dependency was specific to the module. TBH with csv it is fine from my point of view to have it in the top dependencies as this will be also used in other modules.

**Comment by ruflin** [comment]
@hartfordfive Thanks for all the work on this one. I think there are still parts we can tweak in the field naming but for all the other modules we also had here multiple iteration. I left a few very minor comments. The most important one for me is that we mark it experimental. See here https://github.com/elastic/beats/blob/master/metricbeat/module/beats/filebeat/filebeat.go#L26 as an example. This will allow us to change fields and structure with follow up PR's without someone expecting that this is already stable. From my point of view it would be ok to merge it soonish and then go through the hole module again and tweak it in a separate PR. But lets also get @andrewkroh view on this.

File: vendor/vendor.json (hunk 25 above)
Comment (hartfordfive) [inline vendor/vendor.json]: Sorry, I actually missed that file.  The tool used was [govendor](https://github.com/kardianos/govendor).  If you guys have another standard way of bringing in vendor dependencies, let me know and I'll do that instead.

File: vendor/github.com/gocarina/gocsv/README.md (hunk 26 above)
Comment (hartfordfive) [inline vendor/github.com/gocarina/gocsv/README.md]: I'm not exactly sure what I should change in this case?

File: metricbeat/etc/beat.yml (hunk 24 above)
Comment (hartfordfive) [inline metricbeat/etc/beat.yml]: I believe I got these changes when merging in the latest master the other day into my branch.

**Comment by hartfordfive** [comment]
@ruflin I've added the experimental warning messages as suggested.

File: metricbeat/module/haproxy/stat/stat.go [hunk 27]
```diff
@@ -65,13 +67,13 @@ func (m *MetricSet) Fetch() ([]common.MapStr, error) {
 
 	hapc, err := haproxy.NewHaproxyClient(m.statsAddr)
 	if err != nil {
-		return nil, fmt.Errorf(fmt.Sprintf("HAProxy Client error: %s", err))
+		return nil, fmt.Errorf("HAProxy Client error: %s", err)
```
Comment (ruflin) [inline metricbeat/module/haproxy/stat/stat.go]: errors.New would also work here. But that is a minor detail which can be tackled later.

File: metricbeat/etc/beat.yml (hunk 24 above)
Comment (ruflin) [inline metricbeat/etc/beat.yml]: Does not seem to be in master: https://github.com/elastic/beats/blob/master/metricbeat/etc/beat.yml Could you run `git checkout master metricbeat/etc/beat.yml` ?

File: metricbeat/metricbeat.yml (hunk 12 above)
Comment (ruflin) [inline metricbeat/metricbeat.yml]: @hartfordfive Can you please also revert this change?

File: metricbeat/metricbeat.yml (hunk 13 above)
Comment (ruflin) [inline metricbeat/metricbeat.yml]: @hartfordfive This also should be set to false by default.

**Comment by ruflin** [comment]
@hartfordfive Thanks. Please check again my most recent comments above. I think it is important that we don't change our global defaults. Could be that these change happened sometimes during merging.

**Comment by ruflin** [comment]
@hartfordfive Merged. Thanks a lot for your contribution. I opened a follow up PR here to do some fixes like fixing the doc build: https://github.com/elastic/beats/pull/2486

I also opened the github issue here to track some improvements that we have to do: https://github.com/elastic/beats/issues/2485
